### PR TITLE
Implement initial port support for PostgreSQL driver

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue #18578 cc: @irbian

### Summary of Changes

Add the port option to the PostgreSQL driver

### Steps to reproduce the issue

Have a connection on a port different than the default and try to make a query on it
```php
        $options = array();
        
        $options['driver']   = 'postgresql';        
        $options['host']     = '192.192.192.192';
        $options['port'] = '15432'; 
        $options['user']     = 'myuser';       
        $options['password'] = 'mypassword'; 
        $options['database'] = 'MYDATABASE';
        $options['prefix']   = '';             
        
        $dbo =  JDatabaseDriver::getInstance( $options );
```
### Expected result

I expect to work

### Actual result

"Error connecting to PGSQL database."

### Open Points

- [x] Do we need want this kind of port detection: https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/database/driver/mysqli.php#L105-L154 like we have it for mysql?

### CODE REVIEW

**Comment by mbabker** [comment]
FWIW the Framework handled this a bit differently, see https://github.com/joomla/joomla-framework/pull/141 for the PR adding it (implementation is basically still the same now).

And no, I don't think you need the same type of detection as the MySQLi driver.  That honestly seems really verbose for reasons I don't understand without getting into the commit logs.

**Comment by zero-24** [comment]
Thanks @mbabker i have just implemented the port detection from the framework! @irbian please feel free to test this one.

**Review by mbabker — APPROVED** [review]

File: libraries/joomla/database/driver/postgresql.php [hunk 1]
```diff
@@ -119,7 +155,7 @@ public function connect()
 			$dsn .= "host={$this->options['host']} ";
 		}
 
-		$dsn .= "dbname={$this->options['database']} user={$this->options['user']} password={$this->options['password']}";
+		$dsn .= "dbname={$this->options['database']} port={$this->options['port']} user={$this->options['user']} password={$this->options['password']}";
```
Comment (csthomas) [inline libraries/joomla/database/driver/postgresql.php]: I'm not sure but if you always adding port=... then you will always force connection by TCP/IP and remove option to connect by unix socket, which usual faster.

File: libraries/joomla/database/driver/postgresql.php [hunk 2]
```diff
@@ -111,6 +112,41 @@ public function connect()
...
+		/*
+		 * pg_connect() takes the port as separate argument. Therefore, we
+		 * have to extract it from the host string (if povided).
+		 */
+
+		// Check for empty port
+		if (!$this->options['port'])
+		{
+			// Port is empty or not set via options, check for port annotation (:) in the host string
+			$tmp = substr(strstr($this->options['host'], ':'), 1);
```
Comment (Quy) [inline libraries/joomla/database/driver/postgresql.php:124]: You can do `explode(':', $this->options['host'])`. Then `$tmp[0]` will be the domain and `$tmp[1]` will be the port if provided.

File: libraries/joomla/database/driver/postgresql.php (hunk 1 above)
Comment (zero-24) [reply libraries/joomla/database/driver/postgresql.php]: Ok i'm going to move the port setting up to the host so if the host is not set we try without host &port for a socket Connection

File: libraries/joomla/database/driver/postgresql.php (hunk 2 above)
Comment (zero-24) [reply libraries/joomla/database/driver/postgresql.php:124]: I have used the code from here: https://github.com/joomla-framework/database/blob/master/src/Postgresql/PostgresqlDriver.php so i think we should fix it in the source first can you send a PR with the needed changes?

File: libraries/joomla/database/driver/postgresql.php (hunk 1 above)
Comment (zero-24) [reply libraries/joomla/database/driver/postgresql.php]: Sorry took longer than expected but it is done now :)

**Comment by zero-24** [comment]
What is holding us back from testing and merging this thing or should this just be closed as the interest is gone?

**Comment by mbabker** [comment]
> What is holding us back from testing and merging this thing

Finding the 3 people who actually use Joomla with PostgreSQL support and getting their feedback?

**Comment by alikon** [comment]
i've tested successully on not default port like 5444

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on de3d1047f7306a96c4662b18e60fa5219eb44c39<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19466">issues.joomla.org/tracker/joomla-cms/19466</a>.</sub>

File: libraries/joomla/database/driver/postgresql.php [hunk 3]
```diff
@@ -111,12 +112,47 @@ public function connect()
 			throw new JDatabaseExceptionUnsupported('The pgsql extension for PHP is not installed or enabled.');
 		}
 
+		/*
+		 * pg_connect() takes the port as separate argument. Therefore, we
+		 * have to extract it from the host string (if povided).
```
Comment (Quy) [inline libraries/joomla/database/driver/postgresql.php]: spelling `povided`

Comment (zero-24) [reply libraries/joomla/database/driver/postgresql.php]: fixed by the last commit. thanks.

**Comment by twister65** [comment]
I can not find a port number field in the System -> Global configuration -> Server -> Database Settings form.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19466">issues.joomla.org/tracker/joomla-cms/19466</a>.</sub>

**Comment by Quy** [comment]
@twister65 You can append the port to the Host eg. `192.192.192.192:15432`

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on 2111f2fb089ff6f6a993ea07f0fbb93f38649428<br /><br />Tested successfully with port 5466 and postgresql driver (as well as the already implemented port with pgsql PDO driver).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19466">issues.joomla.org/tracker/joomla-cms/19466</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19466">issues.joomla.org/tracker/joomla-cms/19466</a>.</sub>

**Pull request merged**